### PR TITLE
[actpool] use heap.Remove() instead of heap.Init() in actqueue.go

### LINKED ISSUE #2995: [actpool] use heap.Remove() instead of heap.Init() in `actqueue.go`
**What would you like to be added**:
use heap.Remove() to delete the element 

**Why is this needed**:
https://github.com/iotexproject/iotex-core/blob/master/actpool/actqueue.go#L163
It is unnecessary to reheap when deleting one element from the heap

### PR DESCRIPTION
**Pull request by mmyj**
close #2995 

~~`heap.Remove` is not fit in this case, I use `heap.Fix` instead of it.~~

following is my bench result
```
BenchmarkHeapFixWithInit/heap.Init-8         	 5078799	       233.5 ns/op
BenchmarkHeapFixWithInit/heap.Fix-8          	16654287	        72.61 ns/op
```

### CODE REVIEW

File: actpool/actqueue.go [hunk 1]
```diff
@@ -154,13 +155,18 @@ func (q *actQueue) cleanTimeout() []action.SealedEnvelope {
...
 			q.index[i] = q.index[size-1]
+			removedIndex = i
 			size--
 			continue
 		}
 		i++
 	}
 	q.index = q.index[:size]
-	heap.Init(&q.index)
+	if len(removedFromQueue) == 1 {
```
Comment (Liuhaai) [inline actpool/actqueue.go:165]: If the number of items to be removed is larger than 1, the time complexity is still O(n)

**Comment by Liuhaai** [comment]
Here is my pesudocode for the improvement:
```
for i := 0; i < size; i++ {
		if timeNow.After(q.index[i].deadline) {
			removedFromQueue = append(removedFromQueue, q.items[q.index[i].nonce])
			delete(q.items, q.index[i].nonce)
			removedFromIndex = append(removedFromIndex, q.index[i])
		}
	}
for i := range removedFromIndex {
		heap.Remove(&q.index, removedFromIndex[i].idx)
}
```

**Comment by mmyj** [comment]
> Here is my pesudocode for the improvement:
> 
> ```
> for i := 0; i < size; i++ {
> 		if timeNow.After(q.index[i].deadline) {
> 			removedFromQueue = append(removedFromQueue, q.items[q.index[i].nonce])
> 			delete(q.items, q.index[i].nonce)
> 			removedFromIndex = append(removedFromIndex, q.index[i])
> 		}
> 	}
> for i := range removedFromIndex {
> 		heap.Remove(&q.index, removedFromIndex[i].idx)
> }
> ```

If we have to remove `m` elements, we need to invoke `heap.Remove()` m times. Then the time complexity is `O(mlogn)`. The average time complexity is `O(n/2 * logn)`. I had written a benchmark, the result show that using `heap.Init` is a better solution in the case. And using `heap.Remove` is better when only have 10% elements need to remove.

Should we add a judgment logic like this
```
if len(removedFromIndex) < size/10 {
  for i := range removedFromIndex {
	heap.Remove(&q.index, removedFromIndex[i].idx)
  }
}else{
  q.index = q.index[:size]
  heap.Init()
}
```

```
func BenchmarkHeapRemove(b *testing.B) {
	const batch = 10
	testIndex := noncePriorityQueue{}
	index := noncePriorityQueue{}
	invalidTime := time.Now()
	validTime := time.Now().Add(10 * time.Minute)
	for k := uint64(1); k <= batch; k++ {
		for j := uint64(0); j < batch; j++ {
			if j < k {
				heap.Push(&testIndex, &nonceWithTTL{nonce: j, deadline: invalidTime})
			} else {
				heap.Push(&testIndex, &nonceWithTTL{nonce: j, deadline: validTime})
			}
		}
		b.ResetTimer()
		b.Run(fmt.Sprintf("heap.Remove-(%d/%d)", k, batch), func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				// init
				index = index[:0]
				for _, nonce := range testIndex {
					nonce2 := *nonce
					index = append(index, &nonce2)
				}
				// algo
				removedNonceList := make([]*nonceWithTTL, 0, batch)
				for _, nonce := range index {
					if invalidTime.Equal(nonce.deadline) {
						removedNonceList = append(removedNonceList, nonce)
					}
				}
				for _, removedNonce := range removedNonceList {
					heap.Remove(&index, removedNonce.idx)
				}
			}
		})
		b.ResetTimer()
		b.Run(fmt.Sprintf("heap.Init-(%d/%d)", k, batch), func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				// init
				index = index[:0]
				for _, nonce := range testIndex {
					nonce2 := *nonce
					index = append(index, &nonce2)
				}
				// algo
				size := index.Len()
				for j := 0; j < size; {
					if invalidTime.Equal(index[j].deadline) {
						index[j] = index[size-1]
						size--
						continue
					}
					j++
				}
				index = index[:size]
				heap.Init(&index)
			}
		})
	}
}
```

this is the result
```
BenchmarkHeapRemove/heap.Remove-(1/20)
BenchmarkHeapRemove/heap.Remove-(1/20)-8         	  890184	      1161 ns/op
BenchmarkHeapRemove/heap.Init-(1/20)
BenchmarkHeapRemove/heap.Init-(1/20)-8           	 1000000	      1175 ns/op
BenchmarkHeapRemove/heap.Remove-(2/20)
BenchmarkHeapRemove/heap.Remove-(2/20)-8         	  482652	      2324 ns/op
BenchmarkHeapRemove/heap.Init-(2/20)
BenchmarkHeapRemove/heap.Init-(2/20)-8           	  487642	      2455 ns/op
BenchmarkHeapRemove/heap.Remove-(3/20)
BenchmarkHeapRemove/heap.Remove-(3/20)-8         	  344175	      3672 ns/op
BenchmarkHeapRemove/heap.Init-(3/20)
BenchmarkHeapRemove/heap.Init-(3/20)-8           	  329853	      3504 ns/op
BenchmarkHeapRemove/heap.Remove-(4/20)
BenchmarkHeapRemove/heap.Remove-(4/20)-8         	  244645	      4707 ns/op
BenchmarkHeapRemove/heap.Init-(4/20)
BenchmarkHeapRemove/heap.Init-(4/20)-8           	  240574	      4715 ns/op
BenchmarkHeapRemove/heap.Remove-(5/20)
BenchmarkHeapRemove/heap.Remove-(5/20)-8         	  187291	      6117 ns/op
BenchmarkHeapRemove/heap.Init-(5/20)
BenchmarkHeapRemove/heap.Init-(5/20)-8           	  204614	      5850 ns/op
BenchmarkHeapRemove/heap.Remove-(6/20)
BenchmarkHeapRemove/heap.Remove-(6/20)-8         	  153465	      7734 ns/op
BenchmarkHeapRemove/heap.Init-(6/20)
BenchmarkHeapRemove/heap.Init-(6/20)-8           	  157945	      7174 ns/op
BenchmarkHeapRemove/heap.Remove-(7/20)
BenchmarkHeapRemove/heap.Remove-(7/20)-8         	  130717	      8912 ns/op
BenchmarkHeapRemove/heap.Init-(7/20)
BenchmarkHeapRemove/heap.Init-(7/20)-8           	  136188	      8163 ns/op
BenchmarkHeapRemove/heap.Remove-(8/20)
BenchmarkHeapRemove/heap.Remove-(8/20)-8         	  113617	     10826 ns/op
BenchmarkHeapRemove/heap.Init-(8/20)
BenchmarkHeapRemove/heap.Init-(8/20)-8           	  121737	      9972 ns/op
BenchmarkHeapRemove/heap.Remove-(9/20)
BenchmarkHeapRemove/heap.Remove-(9/20)-8         	   91332	     12232 ns/op
BenchmarkHeapRemove/heap.Init-(9/20)
BenchmarkHeapRemove/heap.Init-(9/20)-8           	  111949	     10606 ns/op
BenchmarkHeapRemove/heap.Remove-(10/20)
BenchmarkHeapRemove/heap.Remove-(10/20)-8        	   86325	     14085 ns/op
BenchmarkHeapRemove/heap.Init-(10/20)
BenchmarkHeapRemove/heap.Init-(10/20)-8          	  102498	     12185 ns/op
BenchmarkHeapRemove/heap.Remove-(11/20)
BenchmarkHeapRemove/heap.Remove-(11/20)-8        	   75958	     15225 ns/op
BenchmarkHeapRemove/heap.Init-(11/20)
BenchmarkHeapRemove/heap.Init-(11/20)-8          	   93831	     12879 ns/op
BenchmarkHeapRemove/heap.Remove-(12/20)
BenchmarkHeapRemove/heap.Remove-(12/20)-8        	   70282	     17049 ns/op
BenchmarkHeapRemove/heap.Init-(12/20)
BenchmarkHeapRemove/heap.Init-(12/20)-8          	   84780	     14124 ns/op
BenchmarkHeapRemove/heap.Remove-(13/20)
BenchmarkHeapRemove/heap.Remove-(13/20)-8        	   62737	     19157 ns/op
BenchmarkHeapRemove/heap.Init-(13/20)
BenchmarkHeapRemove/heap.Init-(13/20)-8          	   72120	     15423 ns/op
BenchmarkHeapRemove/heap.Remove-(14/20)
BenchmarkHeapRemove/heap.Remove-(14/20)-8        	   57510	     21181 ns/op
BenchmarkHeapRemove/heap.Init-(14/20)
BenchmarkHeapRemove/heap.Init-(14/20)-8          	   75040	     16506 ns/op
BenchmarkHeapRemove/heap.Remove-(15/20)
BenchmarkHeapRemove/heap.Remove-(15/20)-8        	   52670	     22705 ns/op
BenchmarkHeapRemove/heap.Init-(15/20)
BenchmarkHeapRemove/heap.Init-(15/20)-8          	   66012	     17924 ns/op
BenchmarkHeapRemove/heap.Remove-(16/20)
BenchmarkHeapRemove/heap.Remove-(16/20)-8        	   47487	     24265 ns/op
BenchmarkHeapRemove/heap.Init-(16/20)
BenchmarkHeapRemove/heap.Init-(16/20)-8          	   65028	     18667 ns/op
BenchmarkHeapRemove/heap.Remove-(17/20)
BenchmarkHeapRemove/heap.Remove-(17/20)-8        	   45854	     26140 ns/op
BenchmarkHeapRemove/heap.Init-(17/20)
BenchmarkHeapRemove/heap.Init-(17/20)-8          	   60856	     20190 ns/op
BenchmarkHeapRemove/heap.Remove-(18/20)
BenchmarkHeapRemove/heap.Remove-(18/20)-8        	   39044	     29301 ns/op
BenchmarkHeapRemove/heap.Init-(18/20)
BenchmarkHeapRemove/heap.Init-(18/20)-8          	   57674	     20864 ns/op
BenchmarkHeapRemove/heap.Remove-(19/20)
BenchmarkHeapRemove/heap.Remove-(19/20)-8        	   37363	     30536 ns/op
BenchmarkHeapRemove/heap.Init-(19/20)
BenchmarkHeapRemove/heap.Init-(19/20)-8          	   51982	     22500 ns/op
BenchmarkHeapRemove/heap.Remove-(20/20)
BenchmarkHeapRemove/heap.Remove-(20/20)-8        	   36356	     32983 ns/op
BenchmarkHeapRemove/heap.Init-(20/20)
BenchmarkHeapRemove/heap.Init-(20/20)-8          	   52341	     23240 ns/op
```

**Comment by Liuhaai** [comment]
Thx for the benchmark reports. IMO, O(lgn) should be theoretically slower than O(n). But heap.pop() costs some time in heap.remove(), so it doesn't make perfromance better, even worse.
For code clearity, we can keep the code unchanged. Could you share the benchmark test (the test name `BenchmarkHeapInitAndRemove` is preferred) in this pr? It will give a explaination why heap.init() is used here.

File: actpool/actqueue.go [hunk 2]
```diff
@@ -23,6 +23,7 @@ import (
 )
 
 type nonceWithTTL struct {
+	idx      int
```
Comment (mmyj) [inline actpool/actqueue.go:26]: I add a `idx` property, and it is only use for benchmark `heap.Remove`, is it acceptable?

**Comment by mmyj** [comment]
> Thx for the benchmark reports. It's a surprising finding. IMO, O(lgn) should be theoretically slower than O(n). But heap.pop() might cost some time in heap.remove(), so it doesn't make perfromance better, even worse. For code clearity, we can keep the code unchanged. Could you share the benchmark test (the test name `BenchmarkHeapInitAndRemove` is preferred) in this pr? It will give an explaination why heap.init() is used here.

Done, PTAL.

File: actpool/actqueue.go [hunk 3]
```diff
@@ -31,10 +32,15 @@ type noncePriorityQueue []*nonceWithTTL
...
-func (h noncePriorityQueue) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+func (h noncePriorityQueue) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+	h[i].idx = i
+	h[j].idx = j
+}
 
 func (h *noncePriorityQueue) Push(x interface{}) {
 	if in, ok := x.(*nonceWithTTL); ok {
+		in.idx = h.Len()
```
Comment (Liuhaai) [inline actpool/actqueue.go:43]: `in.idx = len(*h)` Don't use public func inside another public func.

File: actpool/actqueue.go (hunk 2 above)
Comment (Liuhaai) [reply actpool/actqueue.go:26]: it's ok

**Review by Liuhaai — APPROVED** [review]

**Comment by Liuhaai** [comment]
thx for the contribution

File: actpool/actqueue.go (hunk 3 above)
Comment (mmyj) [reply actpool/actqueue.go]: ok, fixed

**Pull request merged**